packages/eslint-plugin/src/rules/no-for-in-array.ts

```diff
@@ -45,3 +45,20 @@ export default createRule({
     };
   },
 });
+
+function isTypeArrayTypeOrUnionOfArrayTypes(
+  type: ts.Type,
+  checker: ts.TypeChecker,
+): boolean {
+  for (const t of tsutils.unionTypeParts(type)) {
+    if (tsutils.isTypeFlagSet(t, ts.TypeFlags.Undefined | ts.TypeFlags.Null)) {
+      continue;
+    }
+
+    if (!checker.isArrayType(t)) {
+      return false;
+    }
+  }
+
+  return true;
```

Comment:
(optional) This fixes a sort-of bug where a purely nullish type is considered an array type... "sort-of", since TS reports an error for the cases where the rule would reach this bug. 

```suggestion
  const nonNullishParts = tsutils
    .unionTypeParts(type)
    .filter(
      t =>
        !tsutils.isTypeFlagSet(t, ts.TypeFlags.Undefined | ts.TypeFlags.Null),
    );
  return (
    nonNullishParts.length > 0 &&
    nonNullishParts.every(t => checker.isArrayType(t))
  );
```

Reply:
(test case)

[Playground](https://deploy-preview-10535--typescript-eslint.netlify.app/play#ts=5.7.2&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3igrhCAlmgBYBc2eE8APvDiqAGYEojADcAUAPTfwACGNAFoQADwAO4DKJgwkMTowXwAFMnRYA1vBYV8RYgEp4Ab3gBfIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDMPpbmtAIbRoQwuigB3EU0jgwAXxCKgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

```ts
declare const nullish: null | undefined;
// @ts-expect-error
for (const k in nullish) { }
```